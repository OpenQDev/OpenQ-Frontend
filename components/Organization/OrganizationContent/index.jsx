import React from 'react';

// Custom
import BountyList from '../../BountyList';
import RepoCard from '../RepoCard';

const OrganizationContent = ({ repositories, organizationData, paginationObj }) => {
  return (
    <div className='max-w-[960px] w-full md:basis-3/4 md:shrink'>
      <h2 className='text-primary w-full mb-2'>Active Repos</h2>
      <div className='grid md:grid-cols-[1fr_1fr] gap-4 pb-5'>
        {repositories.map((repository, index) => (
          <RepoCard key={index} repository={repository} organizationData={organizationData} />
        ))}
      </div>

      <h2 className='text-primary w-full mb-2'>Smart Contracts</h2>
      <BountyList paginationObj={paginationObj} contractToggle={true} types={['0', '1', '2', '3']} />
    </div>
  );
};

export default OrganizationContent;
