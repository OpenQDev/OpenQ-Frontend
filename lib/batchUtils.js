import { ethers } from 'ethers';
import _ from 'lodash';
import tierWinnerTransactionTemplate from '../constants/tierWinnerTransactionTemplate.json';
import supportingDocumentsCompleteTransactionTemplate from '../constants/supportingDocumentsCompleteTransactionTemplate.json';

const convertCsvToJson = (csvData) => {
  const headers = csvData[0];
  const rows = csvData.slice(1);
  const jsonData = rows.map((row) => {
    const jsonObject = {};
    row.forEach((cell, index) => {
      jsonObject[headers[index]] = cell;
    });
    return jsonObject;
  });
  return jsonData;
};

const convertPayoutScheduleToBigInt = (payoutSchedule, decimals) => {
  const payoutScheduleParsed = payoutSchedule && JSON.parse(payoutSchedule);

  const newPayoutSchedule = payoutScheduleParsed.map((tierVolume) => {
    let formattedVolume = tierVolume * 10 ** decimals;
    return ethers.BigNumber.from(formattedVolume.toLocaleString('fullwide', { useGrouping: false }));
  });

  return newPayoutSchedule;
};

const abiEncodeTieredFixed = (initializationData) => {
  let abiCoder = new ethers.utils.AbiCoder();
  const initializationSchema = ['uint256[]', 'address', 'bool', 'bool', 'bool', 'string', 'string', 'string'];

  for (let i = 0; i < initializationData.length; i++) {
    if (initializationData[i] === 'TRUE') {
      initializationData[i] = 1;
    } else if (initializationData[i] === 'FALSE') {
      initializationData[i] = 0;
    }
  }

  const tieredFixedEncoded = abiCoder.encode(initializationSchema, initializationData);

  let tieredFixed = [3, `\"${tieredFixedEncoded}\"`];

  return tieredFixed;
};

const getUnclaimedTierWithVolume = (
  bountyId,
  payoutSchedule,
  tierWinners,
  tiersClaimedPreviouslyInBatch,
  bigNumberTierVolume
) => {
  const indexes = [];

  payoutSchedule.forEach((value, index) => {
    if (ethers.BigNumber.from(value).eq(bigNumberTierVolume)) {
      indexes.push(index);
    }
  });

  let filteredIndexesOnTierWinner = indexes;
  if (tierWinners.length > 0) {
    filteredIndexesOnTierWinner = indexes.filter((index) => {
      return tierWinners[index] === '';
    });
  }

  const filteredIndexesOnTierWinnerPrevious = filteredIndexesOnTierWinner.filter((index) => {
    let previousTiers = [];
    if (tiersClaimedPreviouslyInBatch[bountyId]) {
      previousTiers = tiersClaimedPreviouslyInBatch[bountyId];
    } else {
      previousTiers = [];
    }

    return !previousTiers.includes(index);
  });

  return filteredIndexesOnTierWinnerPrevious.length == 0 ? null : filteredIndexesOnTierWinnerPrevious[0];
};

const getSetSupportingDocumentsCompleteTransactions = async (
  jsonData,
  openQAddress,
  loadGithubData,
  loadGithubDataUser,
  loadOnChainBounty
) => {
  return new Promise(async (resolve, reject) => {
    const transactions = [];

    for (const transactionData of jsonData) {
      const { githubIssueUrl, winnerGithubProfileUrl } = transactionData;

      try {
        const { bountyId } = await loadGithubData(githubIssueUrl);
        const { tierWinners } = await loadOnChainBounty(bountyId);

        let userId;
        try {
          userId = await loadGithubDataUser(winnerGithubProfileUrl);
        } catch (error) {
          return reject(`Failed to load user for URL ${winnerGithubProfileUrl}. Is the case correct?`);
        }

        const tier = tierWinners.findIndex((tierWinner) => tierWinner === userId);

        if (tier == -1) {
          return reject(`User ${winnerGithubProfileUrl} for bounty ${githubIssueUrl} is not in ${tierWinners}.`);
        }

        const abiCoder = new ethers.utils.AbiCoder();
        const _data = abiCoder.encode(['uint256', 'bool'], [tier, true]);

        const supportingDocsBatchTransactionCopy = _.cloneDeep(supportingDocumentsCompleteTransactionTemplate);

        supportingDocsBatchTransactionCopy.to = openQAddress;

        supportingDocsBatchTransactionCopy.contractInputsValues._bountyId = bountyId;
        supportingDocsBatchTransactionCopy.contractInputsValues._data = _data;

        transactions.push(supportingDocsBatchTransactionCopy);
      } catch (err) {
        return reject(err);
      }
    }

    return resolve(transactions);
  });
};

const mapIdToArray = (map, id, element) => {
  // Check if the id already exists in the map
  if (map[id]) {
    // If it exists, push the element onto the array at that index
    map[id].push(element);
  } else {
    // If it doesn't exist, create a new array with the element and map it to the id
    map[id] = [element];
  }

  return map;
};

const getTierWinnerTransactions = async (
  jsonData,
  openQAddress,
  loadGithubData,
  loadGithubDataUser,
  loadOnChainBounty,
  tokenClient
) => {
  return new Promise(async (resolve, reject) => {
    const transactions = [];
    const tiersClaimedPreviouslyInBatchForBounty = {};

    for (const transactionData of jsonData) {
      const { githubIssueUrl, tierAmount, winnerGithubProfileUrl } = transactionData;

      try {
        const { bountyId } = await loadGithubData(githubIssueUrl);

        const bounty = await loadOnChainBounty(bountyId);

        const { payoutSchedule, payoutTokenAddress, tierWinners } = bounty;

        const { decimals } = await tokenClient.getToken(ethers.utils.getAddress(payoutTokenAddress));

        let formattedVolume = tierAmount * 10 ** decimals;

        const bigNumberTierVolume = ethers.BigNumber.from(
          formattedVolume.toLocaleString('fullwide', { useGrouping: false })
        );

        const tier = getUnclaimedTierWithVolume(
          bountyId,
          payoutSchedule,
          tierWinners ? tierWinners : [],
          tiersClaimedPreviouslyInBatchForBounty,
          bigNumberTierVolume
        );

        if (tier == null) {
          return reject(
            `User ${winnerGithubProfileUrl} for bounty ${githubIssueUrl} has no available tiers for ${tierAmount}.\ Tier winners are ${tierWinners}}. Payout schedule is ${payoutSchedule}`
          );
        }

        mapIdToArray(tiersClaimedPreviouslyInBatchForBounty, bountyId, tier);

        let userId;
        try {
          userId = await loadGithubDataUser(winnerGithubProfileUrl, tier);
        } catch (error) {
          return reject(`Failed to load user for URL ${winnerGithubProfileUrl}. Is the case correct?`);
        }

        const tierWinnerTransactionTemplateCopy = _.cloneDeep(tierWinnerTransactionTemplate);

        tierWinnerTransactionTemplateCopy.to = openQAddress;

        tierWinnerTransactionTemplateCopy.contractInputsValues._bountyId = bountyId;
        tierWinnerTransactionTemplateCopy.contractInputsValues._tier = tier.toString();
        tierWinnerTransactionTemplateCopy.contractInputsValues._winner = userId;

        transactions.push(tierWinnerTransactionTemplateCopy);
      } catch (err) {
        return reject(err);
      }
    }

    return resolve(transactions);
  });
};

export {
  convertPayoutScheduleToBigInt,
  convertCsvToJson,
  abiEncodeTieredFixed,
  getUnclaimedTierWithVolume,
  getTierWinnerTransactions,
  getSetSupportingDocumentsCompleteTransactions,
};
