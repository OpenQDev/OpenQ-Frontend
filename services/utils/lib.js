import { ethers } from 'ethers';

export const getNonBlacklisted = async (appState, repoName, org, limit) => {
  const { repoPrs, totalCount } = await appState.githubRepository.getPrs(org, repoName, limit || 3);

  const prs = await appState.openQPrismaClient.getSubmissions();
  const blacklistedPrIds = prs.filter((pr) => pr.blacklisted).map((pr) => pr.id);

  return { nonBlacklisted: repoPrs.filter((pr) => !blacklistedPrIds.includes(pr.id)), totalCount };
};

export const shortenAddress = (address) => {
  if (!address) {
    return '';
  }
  return `${address.slice(0, 4)}...${address.slice(38)}`;
};

export const parseVolume = (volume) => {
  const numberRegex = /^(\d+)?(\.)?(\d+)?$/;
  if (numberRegex.test(volume) || volume === '' || volume === '.') {
    return volume.match(numberRegex)[0];
  } else {
    return null;
  }
};
export const listWordsWithAnd = (words) => {
  if (words.length === 0) return '';
  if (words.length === 1) {
    return words[0];
  }
  if (words.length === 2) {
    return `${words[0]} and ${words[1]}`;
  }
  return `${words.slice(0, -1).join(', ')}, and ${words[words.length - 1]}`;
};

export const capitalize = (word) => {
  if (!word) return '';
  return word[0].toUpperCase() + word.substring(1);
};
export const valueToDisplay = (value) => {
  switch (value) {
    case 'invoicingName':
      return 'Invoicing Name';
    case 'invoicingEmail':
      return 'Invoicing Email';
    case 'streetAddress':
      return 'Billing Address';
    case 'postalCode':
      return 'Postal Code';
    case 'phoneNumber':
      return 'Phone Number';
    case 'invoiceNumber':
      return 'Invoice Number';
    case 'taxId':
      return 'Tax ID';
    case 'vatNumber':
      return 'VAT Number';
    case 'vatRate':
      return 'VAT Rate';
    case 'billingName':
      return 'Billing Name';
    default:
      return capitalize(value);
  }
};

export const getBigNumberVol = (volume, token) => {
  const volumeFloat = parseFloat(volume) || 0;
  const volumeInWei = volumeFloat * 10 ** token.decimals;

  return ethers.BigNumber.from(volumeInWei.toLocaleString('fullwide', { useGrouping: false }));
};

export const checkHackathonDates = (startDate, endDate, today) => {
  const start = startDate && new Date(startDate);
  const end = endDate && new Date(endDate);
  if ((start && start < today) || (end && end < today) || (start && end && end < start)) {
    return false;
  }
  return true;
};
const checkPrUsed = (pr, bounty) => {
  return bounty.claims?.some((claim) => claim.claimantAsset === pr.source.url);
};
const checkTierClaimed = (bounty, index) => {
  return bounty.claims?.some((claim) => claim.tier === index.toString());
};

export const checkFixedAndSplit = (bounty, currentUser) => {
  if (
    bounty.status == '0' &&
    bounty?.prs?.some((pr) => pr.source.merged && pr.source.author.id === currentUser && checkPrUsed(pr, bounty))
  )
    return { status: 'Claimed' };
  if (bounty.status == '0' && bounty?.prs?.some((pr) => pr.source.merged && pr.source.author.id === currentUser)) {
    return { status: 'Claimable' };
  }
  return { status: null };
};

export const getWinningPrOfUser = (bounty, currentUser) => {
  if (bounty?.tierWinners?.some((winner, index) => winner === currentUser && checkTierClaimed(bounty, index))) {
    const winningPr = bounty?.prs?.find((pr) => pr.source.author.id === currentUser);

    return winningPr.source;
  }
};

export const checkTiered = (bounty, currentUser) => {
  if (bounty?.tierWinners?.some((winner, index) => winner === currentUser && checkTierClaimed(bounty, index))) {
    return { status: 'Claimed' };
  }

  if (bounty?.tierWinners?.some((winner) => winner === currentUser)) {
    return { status: 'Claimable' };
  }
  return { status: null };
};

export const checkClaimable = (bounty, currentUser) => {
  bounty.bountyType === '0';
  switch (bounty.bountyType) {
    case '0': {
      return checkFixedAndSplit(bounty, currentUser);
    }
    case '1': {
      return checkFixedAndSplit(bounty, currentUser);
    }
    case '2': {
      return checkTiered(bounty, currentUser);
    }
    case '3': {
      return checkTiered(bounty, currentUser);
    }
    default:
      return { status: null };
  }
};

export const getBountyMarker = (bounty, openQClient, githubId, checkClaimableImpl = checkClaimable) => {
  if (bounty.closed) return { status: 'Closed', colour: 'bg-danger', fill: 'fill-danger' };
  const { status } = checkClaimableImpl(bounty, githubId, openQClient);
  if (status === 'Claimable') {
    return {
      status: 'Claim Available',
      colour: 'bg-closed',
      fill: 'fill-closed',
    };
  }
  if (status === 'Claimed') {
    return {
      status: 'Claimed',
      colour: 'bg-closed',
      fill: 'fill-closed',
    };
  }
  if (bounty.bountyType === '0') {
    if (bounty.assignees[0]) {
      return {
        status: 'In Progress',
        colour: 'bg-yellow-500 text-black fill-black',
        fill: 'fill-yellow-500',
      };
    }
    return {
      status: 'Ready for Work',
      colour: 'bg-green',
      fill: 'fill-green',
    };
  } else {
    // for split price and contests, closed when status is 1
    if (bounty.status == '1') {
      return { status: 'Closed', colour: 'bg-danger', fill: 'fill-danger' };
    } else {
      return { status: 'Open', colour: 'bg-green', fill: 'fill-green' };
    }
  }
};

export const getPlural = (count) => {
  if (count > 1 || count === 0) {
    return 's';
  }
  return '';
};

export const getBool = (value) => {
  if (value === 'Yes') {
    return true;
  }
  return false;
};
export const reverseBool = (value) => {
  if (value === true) {
    return 'Yes';
  }
  return 'No';
};

export const getW8Approved = (bounty, accountData) => {
  if (!bounty.supportingDocumentsCompleted) return bounty.supportingDocumentsCompleted;
  if (bounty.bountyType === '2' || bounty.bountyType === '3') {
    return bounty.supportingDocumentsCompleted[bounty.tierWinners?.indexOf(accountData.github)];
  } else {
    return bounty.supportingDocumentsCompleted;
  }
};

export const isContest = (bounty) => {
  return bounty.bountyType === '2' || bounty.bountyType === '3';
};

export const isEveryValueNotNull = (obj) => {
  if (!obj) return false;
  const { kyc, w8Form, githubHasWallet, invoice } = obj;
  return kyc && w8Form && githubHasWallet && invoice;
};
export const formatVolume = (tierVolume, token) => {
  let bigNumberVolume = ethers.BigNumber.from(tierVolume.toString());
  let decimals = parseInt(token?.decimals) || 18;
  let formattedVolume = ethers.utils.formatUnits(bigNumberVolume, decimals);
  return formattedVolume;
};

export const fetchRequestsWithServiceArg = async (appState, identity, oldCursor, batch) => {
  const { userId, githubId, email } = identity;
  const userOffChainData = await appState.openQPrismaClient.getUserRequests(
    {
      id: userId,
      github: githubId,
      email: email,
    },
    {
      bountiesCursor: oldCursor,
      bountiesLimit: batch,
    }
  );
  const createdBounties = userOffChainData.createdBounties.bountyConnection.nodes.filter((bounty) => {
    return bounty.requests;
  });
  const requests = createdBounties
    .map((bounty) => {
      const requests = bounty.requests.nodes.reduce((accum, request) => {
        const user = accum.find(
          (earlierRequest) => earlierRequest.request.requestingUser.id === request.requestingUser.id
        );

        if (!user) {
          return accum.concat({ request, bounty });
        } else return accum;
      }, []);
      return requests;
    })
    .flat();
  return {
    nodes: requests,
    cursor: userOffChainData.createdBounties.bountyConnection.cursor,
    complete: createdBounties.length !== batch,
  };
};

export const fetchBountiesWithServiceArg = async (appState, oldCursor, batch, ordering, filters) => {
  let { sortOrder, field } = ordering;
  if (!sortOrder) {
    sortOrder = 'desc';
  }
  if (!field) {
    field = 'createdAt';
  }
  const { types, organizationId, repositoryId } = filters;

  try {
    let [fullBounties, cursor, complete] = await appState.utils.fetchBounties(
      appState,
      batch,
      types,
      sortOrder,
      field,
      oldCursor,
      organizationId,
      null,
      null,
      repositoryId
    );
    return {
      nodes: fullBounties,
      cursor,
      complete,
    };
  } catch (err) {
    appState.logger.error(err);
    return { nodes: [], cursor: null, complete: true };
  }
};

export const getReadyText = (isContest) => {
  if (isContest) {
    return 'Ready to Hack';
  } else return 'Ready for Work';
};
export const isOnlyContest = (types) => {
  const includesContest = types.includes('2') || types.includes('3');
  const includesNonContest = types.includes('0') || types.includes('1');
  return includesContest && !includesNonContest;
};

export const fetchRepositories = async (appState, variables) => {
  try {
    const repositories = await appState.openQPrismaClient.getRepositories(variables);

    const checkBlacklisted = (repository) => {
      const organizationBlacklisted = repository.organization.blacklisted;
      const nonBlacklistedBounties = repository.bounties.nodes.filter((bounty) => !bounty.blacklisted);
      const bountiesBlacklisted = nonBlacklistedBounties?.length === 0;
      return !organizationBlacklisted && !bountiesBlacklisted;
    };
    const repositoryIds = repositories
      .filter((repository) => checkBlacklisted(repository))
      .map((repository) => repository.id);
    const githubRepositories = await appState.githubRepository.fetchReposByIds([...repositoryIds]);
    return githubRepositories;
  } catch (err) {
    appState.logger.error(err);
    return [];
  }
};

export const getBountyTypeName = (type) => {
  const bountyType = type?.toString();
  switch (bountyType) {
    case '0':
      return 'Fixed Price';
    case '1':
      return 'Split Price';
    case '2':
      return 'Contest';
    case '3':
      return 'Hackathon';
    default:
      return 'Unknown';
  }
};
export const getTypeFromCategory = (category) => {
  switch (category) {
    case 'Fixed Price':
      return 0;
    case 'Split Price':
      return 1;
    case 'Hackathon':
      return 3;
    default:
      return null;
  }
};
export const needsOrgData = (accountData) => {
  const orgAccountKeys = ['company', 'city', 'streetAddress', 'province', 'country', 'invoicingEmail'];
  const neededAccountData = orgAccountKeys.filter((key) => {
    return !accountData[key];
  });
  return neededAccountData.length > 0;
};
export const needsFreelancerData = (accountData) => {
  const accountKeys = [
    'company',
    'billingName',
    'city',
    'streetAddress',
    'postalCode',
    'country',
    'phoneNumber',
    'province',
    'invoicingEmail',
    'invoiceNumber',
    'taxId',
    'vatNumber',
    'vatRate',
  ];
  const neededAccountData = accountKeys.filter((key) => {
    return !accountData[key] || accountData[key] == 0;
  });
  return neededAccountData.length > 0;
};
export const handleDispatch = (e, type, dispatchFunc) => {
  const dispatch = {
    type,
    payload: e.target.value,
  };
  dispatchFunc(dispatch);
};
export const updateHackathonState = async (hackathonState, appState, setCreateHackathonResponse, push) => {
  const {
    repositoryUrl,
    startDate,
    endDate,
    city,
    eventOrganizer,
    isIrl,
    timezone,
    topic,
    website,
    contactEmail,
    twitter,
    discord,
    telegram,
    description,
    slack,
    registrationDeadline,
    eventName,
    proAccountId,
    isDraft,
  } = hackathonState;
  const forwardedDraft = isDraft ? true : false;
  const ownerRegex = /github.com\/([a-zA-Z0-9-]+)\/([a-zA-Z0-9-]+)/;
  const owner = ownerRegex.exec(repositoryUrl)?.[1];
  const name = ownerRegex.exec(repositoryUrl)?.[2];
  try {
    setCreateHackathonResponse('PENDING');
    const githubRepository = await appState.githubRepository.fetchRepoByName(owner, name, []);
    const repositoryId = githubRepository.id;
    const organizationId = githubRepository.owner.id;
    const variables = {
      proAccountId,
      repositoryId,
      startDate,
      endDate,
      organizationId,
      isContest: true,
      isDraft: forwardedDraft,
      city,
      isIrl,
      timezone,
      eventOrganizer,
      repositoryUrl,
      topic,
      website,
      contactEmail,
      twitter,
      discord,
      telegram,
      slack,
      registrationDeadline,
      description,
      eventName,
    };
    await appState.openQPrismaClient.updateRepositoryAsContest(variables);
    setCreateHackathonResponse('SUCCESS');
    push();
  } catch (e) {
    setCreateHackathonResponse('ERROR');
  }
};
export const shortenString = (str, maxLen) => {
  if (str.length > maxLen) {
    return str.substring(0, maxLen) + '...';
  }
  return str;
};
