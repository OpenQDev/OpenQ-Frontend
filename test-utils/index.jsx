// Third party
import React from 'react';
// Cusotm
import { render, renderHook } from '@testing-library/react';
import { cleanup } from '@testing-library/react';

import StoreProvider from '../store/Store/TestStoreProvider';
import AuthProvider from '../store/AuthStore/TestAuthProvider';
import FundProvider from '../components/FundBounty/FundStore/TestFundProvider';
import TokenProvider from '../components/TokenSelection/TokenStore/TokenProvider';

// Add in any providers here if necessary:
// test-utils.js
// (ReduxProvider, ThemeProvider, etc)
afterEach(() => {
  cleanup();
});

const customRender = (ui, options = {}, storeProps, authProps) => {
  const IntersectionObserverMock = vi.fn(() => ({
    disconnect: vi.fn(),
    observe: vi.fn(),
    takeRecords: vi.fn(),
    unobserve: vi.fn(),
  }));
  vi.stubGlobal('IntersectionObserver', IntersectionObserverMock);
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <TokenProvider>{children}</TokenProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return render(ui, { wrapper: Providers, ...options });
};
const customHookRender = (hook, options = {}, storeProps, authProps, fundProps) => {
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <FundProvider StoreProps={fundProps}>
            <TokenProvider>{children}</TokenProvider>
          </FundProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return renderHook(hook, { wrapper: Providers, ...options });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, customHookRender as hookRender };
