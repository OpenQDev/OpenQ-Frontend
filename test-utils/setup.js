import { expect, beforeEach } from 'vitest';
import { cleanup } from '@testing-library/react';
import matchers from '@testing-library/jest-dom/matchers';

expect.extend(matchers);
beforeEach(() => {
  cleanup();
});
process.on('unhandledRejection', (reason) => {
  // eslint-disable-next-line no-console
  console.log(`FAILED TO HANDLE PROMISE REJECTION`);
  throw reason;
});
