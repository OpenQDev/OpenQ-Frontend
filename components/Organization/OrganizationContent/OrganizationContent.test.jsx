/**
 * @vi-environment jsdom
 */
import React from 'react';
import userEvent from '@testing-library/user-event';
import { getReadyText, isOnlyContest } from '../../../services/utils/lib';

import { render, screen } from '../../../test-utils';
import OrganizationContent from '.';
import ShallowRenderer from 'react-test-renderer/shallow';

import nextRouter from 'next/router';

const orgBounties = [
  {
    id: 'I_kwDOCHE8585AYvGo',
    title: 'How to estimate transaction fees for cancelling orders on opensea?',
    body: 'I am a little confused on how to find out how much it would cost to cancel a bid using the SDK. Current gwei price is arround 80 and opensea gives me a 30 USD transaction fee on medium when trying to cancel from their website.\r\n\r\nHow do they calculate these fees? Any way to calculate the cost using their SDK before sending the orders in?',
    url: 'https://github.com/ProjectOpenSea/opensea-js/issues/286',
    languages: [
      { __typename: 'Language', name: 'TypeScript', color: '#3178c6' },
      { __typename: 'Language', name: 'Shell', color: '#89e051' },
      { __typename: 'Language', name: 'JavaScript', color: '#f1e05a' },
    ],
    repoName: 'opensea-js',
    owner: 'ProjectOpenSea',
    avatarUrl: 'https://avatars.githubusercontent.com/u/34966464?v=4',
    labels: [
      { __typename: 'Label', name: 'dev-documentation', color: 'c2e0c6' },
      { __typename: 'Label', name: 'dev-sdk-bug', color: '598E75' },
    ],
    createdAt: '2021-12-14T20:52:29Z',
    closed: false,
    bodyHTML:
      '<p dir="auto">I am a little confused on how to find out how much it would cost to cancel a bid using the SDK. Current gwei price is arround 80 and opensea gives me a 30 USD transaction fee on medium when trying to cancel from their website.</p>\n<p dir="auto">How do they calculate these fees? Any way to calculate the cost using their SDK before sending the orders in?</p>',
    titleHTML: 'How to estimate transaction fees for cancelling orders on opensea?',
    assignees: [],
    number: 286,
    repoUrl: 'https://github.com/ProjectOpenSea/opensea-js',
    repoDescription:
      'JavaScript SDK for the OpenSea marketplace. Let your users buy or sell cryptogoods on your own site!',
    prs: [],
    __typename: 'Bounty',
    bountyAddress: '0x033488800ae672726c34620d4bd817e1590d4cdc',
    bountyType: '0',
    bountyId: 'I_kwDOCHE8585AYvGo',
    bountyMintTime: '1661767968',
    bountyClosedTime: null,
    claimedTransactionHash: null,
    fundingGoalTokenAddress: '0x0000000000000000000000000000000000000000',
    fundingGoalVolume: '0',
    status: '0',
    deposits: [],
    issuer: { __typename: 'User', id: '0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266' },
    bountyTokenBalances: [],
    refunds: [],
    payouts: [],
    tvl: 0,
    address: '0x033488800Ae672726c34620D4Bd817E1590d4cDc',
    blacklisted: false,
    category: 'prime',
  },
];

const repositories = [
  {
    name: 'bert-japanese',
    languages: [
      { __typename: 'Language', name: 'Dockerfile' },
      { __typename: 'Language', name: 'Python' },
      { __typename: 'Language', name: 'Shell' },
      { __typename: 'Language', name: 'Jupyter Notebook' },
    ],
  },
];

describe('OrganizationContent', () => {
  beforeEach(() => {
    const observe = vi.fn();
    const disconnect = vi.fn();

    nextRouter.useRouter = vi.fn();
    nextRouter.useRouter.mockImplementation(() => ({
      query: { type: null },

      prefetch: vi.fn(() => {
        return { catch: vi.fn };
      }),
    }));

    window.IntersectionObserver = vi.fn(() => ({
      observe,
      disconnect,
    }));
  });

  it('should render match DOM Snapshot', () => {
    const shallow = new ShallowRenderer();
    shallow.render(
      <OrganizationContent
        paginationObj={{
          items: orgBounties,
          filters: { searchText: '' },
          fetchFilters: { types: ['1', '2', '3', '4'] },
        }}
        repositories={repositories}
      />
    );
    const tree = shallow.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('should render Org content card for yoheikikuta', async () => {
    // ARRANGE
    const user = userEvent.setup();
    const types = ['1', '2', '3', '4'];
    render(
      <OrganizationContent
        paginationObj={{
          items: orgBounties,
          filters: { searchText: '' },
          fetchFilters: { types },
          ordering: { field: 'createdAt', direction: 'desc' },
          isReady: getReadyText(isOnlyContest(types)),
        }}
        repositories={repositories}
      />
    );
    const name = 'opensea';
    // ASSERT

    await user.click(screen.getByText(/all Issues/i));
    const nameRegex = new RegExp(name.slice(0, 3), 'i');
    const title = await screen.findAllByText(nameRegex);
    expect(title[0]).toBeInTheDocument();

    // should not have null or undefined values
    const nullish = [...screen.queryAllByRole(/null/), ...screen.queryAllByRole(/undefined/)];
    expect(nullish).toHaveLength(0);
  });
});
