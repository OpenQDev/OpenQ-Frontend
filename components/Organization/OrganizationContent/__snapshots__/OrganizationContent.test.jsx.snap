// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OrganizationContent > should render match DOM Snapshot 1`] = `
<div
  className="max-w-[960px] w-full md:basis-3/4 md:shrink"
>
  <h2
    className="text-primary w-full mb-2"
  >
    Active Repos
  </h2>
  <div
    className="grid md:grid-cols-[1fr_1fr] gap-4 pb-5"
  >
    <RepoCard
      repository={
        {
          "languages": [
            {
              "__typename": "Language",
              "name": "Dockerfile",
            },
            {
              "__typename": "Language",
              "name": "Python",
            },
            {
              "__typename": "Language",
              "name": "Shell",
            },
            {
              "__typename": "Language",
              "name": "Jupyter Notebook",
            },
          ],
          "name": "bert-japanese",
        }
      }
    />
  </div>
  <h2
    className="text-primary w-full mb-2"
  >
    Smart Contracts
  </h2>
  <BountyList
    contractToggle={true}
    paginationObj={
      {
        "fetchFilters": {
          "types": [
            "1",
            "2",
            "3",
            "4",
          ],
        },
        "filters": {
          "searchText": "",
        },
        "items": [
          {
            "__typename": "Bounty",
            "address": "0x033488800Ae672726c34620D4Bd817E1590d4cDc",
            "assignees": [],
            "avatarUrl": "https://avatars.githubusercontent.com/u/34966464?v=4",
            "blacklisted": false,
            "body": "I am a little confused on how to find out how much it would cost to cancel a bid using the SDK. Current gwei price is arround 80 and opensea gives me a 30 USD transaction fee on medium when trying to cancel from their website.

How do they calculate these fees? Any way to calculate the cost using their SDK before sending the orders in?",
            "bodyHTML": "<p dir=\\"auto\\">I am a little confused on how to find out how much it would cost to cancel a bid using the SDK. Current gwei price is arround 80 and opensea gives me a 30 USD transaction fee on medium when trying to cancel from their website.</p>
<p dir=\\"auto\\">How do they calculate these fees? Any way to calculate the cost using their SDK before sending the orders in?</p>",
            "bountyAddress": "0x033488800ae672726c34620d4bd817e1590d4cdc",
            "bountyClosedTime": null,
            "bountyId": "I_kwDOCHE8585AYvGo",
            "bountyMintTime": "1661767968",
            "bountyTokenBalances": [],
            "bountyType": "0",
            "category": "prime",
            "claimedTransactionHash": null,
            "closed": false,
            "createdAt": "2021-12-14T20:52:29Z",
            "deposits": [],
            "fundingGoalTokenAddress": "0x0000000000000000000000000000000000000000",
            "fundingGoalVolume": "0",
            "id": "I_kwDOCHE8585AYvGo",
            "issuer": {
              "__typename": "User",
              "id": "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266",
            },
            "labels": [
              {
                "__typename": "Label",
                "color": "c2e0c6",
                "name": "dev-documentation",
              },
              {
                "__typename": "Label",
                "color": "598E75",
                "name": "dev-sdk-bug",
              },
            ],
            "languages": [
              {
                "__typename": "Language",
                "color": "#3178c6",
                "name": "TypeScript",
              },
              {
                "__typename": "Language",
                "color": "#89e051",
                "name": "Shell",
              },
              {
                "__typename": "Language",
                "color": "#f1e05a",
                "name": "JavaScript",
              },
            ],
            "number": 286,
            "owner": "ProjectOpenSea",
            "payouts": [],
            "prs": [],
            "refunds": [],
            "repoDescription": "JavaScript SDK for the OpenSea marketplace. Let your users buy or sell cryptogoods on your own site!",
            "repoName": "opensea-js",
            "repoUrl": "https://github.com/ProjectOpenSea/opensea-js",
            "status": "0",
            "title": "How to estimate transaction fees for cancelling orders on opensea?",
            "titleHTML": "How to estimate transaction fees for cancelling orders on opensea?",
            "tvl": 0,
            "url": "https://github.com/ProjectOpenSea/opensea-js/issues/286",
          },
        ],
      }
    }
    types={
      [
        "0",
        "1",
        "2",
        "3",
      ]
    }
  />
</div>
`;
